controllers/cluster_controller.go

```diff
@@ -167,31 +166,34 @@ func (r *ClusterReconciler) Reconcile(req ctrl.Request) (_ ctrl.Result, reterr e
 
 // reconcile handles cluster reconciliation.
 func (r *ClusterReconciler) reconcile(ctx context.Context, cluster *clusterv1.Cluster) (ctrl.Result, error) {
-	logger := r.Log.WithValues("cluster", cluster.Name, "namespace", cluster.Namespace)
-
-	// Call the inner reconciliation methods.
-	reconciliationErrors := []error{
-		r.reconcileInfrastructure(ctx, cluster),
-		r.reconcileControlPlane(ctx, cluster),
-		r.reconcileKubeconfig(ctx, cluster),
-		r.reconcileControlPlaneInitialized(ctx, cluster),
+	phases := []func(context.Context, *clusterv1.Cluster) (ctrl.Result, error){
+		r.reconcileInfrastructure,
+		r.reconcileControlPlane,
+		r.reconcileKubeconfig,
+		r.reconcileControlPlaneInitialized,
 	}
 
-	// Parse the errors, making sure we record if there is a RequeueAfterError.
 	res := ctrl.Result{}
 	errs := []error{}
-	for _, err := range reconciliationErrors {
-		if requeueErr, ok := errors.Cause(err).(capierrors.HasRequeueAfterError); ok {
-			// Only record and log the first RequeueAfterError.
-			if !res.Requeue {
-				res.Requeue = true
-				res.RequeueAfter = requeueErr.GetRequeueAfter()
-				logger.Error(err, "Reconciliation for Cluster asked to requeue")
-			}
-			continue
+	for _, fn := range phases {
+		// Call the inner reconciliation methods.
+		fnResult, err := fn(ctx, cluster)
+		if err != nil {
+			errs = append(errs, err)
 		}
 
-		errs = append(errs, err)
+		// Requeue immediately takes precedence over any other requeue duration.
+		if fnResult.Requeue {
+			res.Requeue = true
+			res.RequeueAfter = time.Duration(0)
+		}
+
+		// If result is empty, or if the new RequeueAfter is lower than the one specified
+		// overwrite the value. This is meant to give precedence for a reconciliation requeuest
+		// that needs to happen more frequently.
+		if res.IsZero() || fnResult.RequeueAfter < res.RequeueAfter {
+			res.RequeueAfter = fnResult.RequeueAfter
+		}
```

Comment:
I'm wondering if it would make sense to have a helper method/function for comparing two different `ctrl.Result`s for this. This may be helpful for replicating the same behavior elsewhere phases are used during reconciliation.

Hacky quick attempt at such a helper function:
```go
func lowestNonZeroResult(i, j ctrl.Result) ctrl.Result {
	switch {
	case i.IsZero():
		return j
	case j.IsZero():
		return i
	case i.Requeue:
		return i
	case j.Requeue:
		return j
	case i.RequeueAfter < j.RequeueAfter:
		return i
	default:
		return j
	}
}
```